operator/pkg/apis/cassandra/v1beta1/webhook.go

```diff
@@ -5,35 +5,109 @@ package v1beta1
 
 import (
 	"errors"
+	"fmt"
 	"k8s.io/apimachinery/pkg/runtime"
+	"reflect"
 	logf "sigs.k8s.io/controller-runtime/pkg/runtime/log"
 	"sigs.k8s.io/controller-runtime/pkg/webhook"
 )
 
 var log = logf.Log.WithName("api")
 
+// Ensure that no values are improperly set
+func ValidateSingleDatacenter(dc CassandraDatacenter) error {
+	// Ensure serverVersion and serverType are compatible
+
+	if dc.Spec.ServerType == "dse" && dc.Spec.ServerVersion != "6.8.0" {
+		return fmt.Errorf("CassandraDatacenter attempted to use unsupported DSE version '%s'",
+			dc.Spec.ServerVersion)
+	}
+
+	if dc.Spec.ServerType == "cassandra" && dc.Spec.ServerVersion != "3.11.6" && dc.Spec.ServerVersion != "4.0.0" {
+		return fmt.Errorf("CassandraDatacenter attempted to use unsupported Cassandra version '%s'",
+			dc.Spec.ServerVersion)
+	}
+
+	return nil
+}
+
+// Ensure that no values are improperly set
+func ValidateDatacenterFieldChanges(oldDc CassandraDatacenter, newDc CassandraDatacenter) error {
+
+	if oldDc.Spec.ClusterName != newDc.Spec.ClusterName {
+		return errors.New("CassandraDatacenter attempted to change ClusterName")
+	}
+
+	if oldDc.Spec.AllowMultipleNodesPerWorker != newDc.Spec.AllowMultipleNodesPerWorker {
+		return errors.New("CassandraDatacenter attempted to change AllowMultipleNodesPerWorker")
+	}
+
+	if oldDc.Spec.SuperuserSecretName != newDc.Spec.SuperuserSecretName {
+		return errors.New("CassandraDatacenter attempted to change SuperuserSecretName")
+	}
+
+	if oldDc.Spec.ServiceAccount != newDc.Spec.ServiceAccount {
+		return errors.New("CassandraDatacenter attempted to change ServiceAccount")
+	}
+
+	// StorageConfig changes are disallowed
+	if !reflect.DeepEqual(oldDc.Spec.StorageConfig, newDc.Spec.StorageConfig) {
+		return fmt.Errorf("CassandraDatacenter attempted to change StorageConfig")
+	}
+
+	// Topology changes - Racks
+	// - Rack Name and Zone changes are disallowed.
+	// - Removing racks is not supported.
+	// - Reordering the rack list is not supported.
+	// - Any new racks must be added to the end of the current rack list.
+
+	if len(oldDc.Spec.Racks) > len(newDc.Spec.Racks) {
```

Comment:
sadly, we have to add the calls to GetRacks for both of these, I think

Reply:
Is this all to support the special case of the user not specifying any racks?

Reply:
that's right